src/block/compress.rs

```diff
@@ -125,88 +128,72 @@ fn token_from_literal_and_match_length(lit_len: usize, duplicate_length: usize)
 /// Counts the number of same bytes in two byte streams.
 /// `input` is the complete input
 /// `cur` is the current position in the input. it will be incremented by the number of matched bytes
-/// `input_dupl` is a pointer back in the input
+/// `source` either the same as input or an external slice
+/// `candidate` is the candidate position in `source`
 ///
-/// The function ignores the last 7bytes (END_OFFSET) in input as this should be literals.
+/// The function ignores the last END_OFFSET bytes in input as this should be literals.
 #[inline]
 #[cfg(feature = "safe-encode")]
-fn count_same_bytes(input: &[u8], candidate: usize, cur: &mut usize) -> usize {
-    let input_dupl = &input[candidate as usize..];
+fn count_same_bytes(input: &[u8], cur: &mut usize, source: &[u8], candidate: usize) -> usize {
+    const USIZE_SIZE: usize = core::mem::size_of::<usize>();
     let cur_slice = &input[*cur..input.len() - END_OFFSET];
+    let cand_slice = &source[candidate as usize..];
 
     let mut num = 0;
-    const USIZE_SIZE: usize = core::mem::size_of::<usize>();
     for (block1, block2) in cur_slice
         .chunks_exact(USIZE_SIZE)
-        .zip(input_dupl.chunks_exact(USIZE_SIZE))
+        .zip(cand_slice.chunks_exact(USIZE_SIZE))
     {
-        let input_block = as_usize_le(block1);
-        let match_block = as_usize_le(block2);
+        let input_block = usize::from_ne_bytes(block1.try_into().unwrap());
+        let match_block = usize::from_ne_bytes(block2.try_into().unwrap());
 
         if input_block == match_block {
             num += USIZE_SIZE;
         } else {
             let diff = input_block ^ match_block;
-            num += get_common_bytes(diff) as usize;
-            break;
+            num += (diff.to_le().trailing_zeros() / 8) as usize;
```

Comment:
I added the `get_common_bytes` function mainly to increase readability, to understand what this code is doing

Reply:
I inlined to make it similar to the other locations dealing with u32/u16 instead (in the unsafe version of this function).

Zero extending big/little-endian u32/u16 to usize to feed to get_common_bytes is very error prone when we're counting zero bits. I think this way strikes a good balance of explicit/readability/uniformity. What do you think?

Reply:
I think it's fine, but a reader not familiar with the code may not be able to understand semantically what the code does, to count the common bytes